### PR TITLE
Enable new configuration other_hcat_clusters in AZ to talk to Metastores in HA mode

### PR DESCRIPTION
**Pull request by herrlich**
The logic that Azkaban uses to get the delegation token from ‘other_hcat_locations’ is trying to connect to all hcat servers in that field, and getting the delegation token from each hcat server. This is could cause delegation token confusion problems. Thus, we decide to introduce a new property field called “other_hcat_clusters”, which can also be set in the “properties” field in the workflow file. This field requires users to group the hcat servers to different clusters, in the form as “cluster1hcat01, cluster1cat02;cluster2hcat01,cluster2hcat02”. In our implementation, this string will be split by semicolon, and each group(cluster) will be traversed until we can get the delegation token from one of the machines in that cluster.

### CODE REVIEW

File: az-core/src/main/java/azkaban/utils/Props.java [hunk 1]
```diff
@@ -430,6 +430,13 @@ public String getString(final String key) {
   }
 
   /**
+   * Returns a list of clusters with the comma as the separator of the value
+   */
```
Comment (beeramsunitha) [inline az-core/src/main/java/azkaban/utils/Props.java]: Provide an example to clarify whats parsed by the method.

Comment (beeramsunitha) [reply az-core/src/main/java/azkaban/utils/Props.java]: Add unit tests to cover different cases.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 2]
```diff
@@ -122,6 +122,7 @@
    * supported, use comma to separate the values, values are case insensitive.
    */
   private static final String EXTRA_HCAT_LOCATION = "other_hcat_location";
+  private static final String EXTRA_HCAT_CLUSTERS = "other_hcat_clusters";
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Update documentation to provide context around this (so the reader knows how it differes from EXTRA_HCAT_LOCATIONS). Also, we should mark EXTRA_HCAT_LOCATIONS as deprecated and ask users to switch to *_CLUSTERS to take advantage of Metastore HA.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 3]
```diff
@@ -525,24 +527,42 @@ public synchronized void prefetchToken(final File tokenFile,
 
         cred.addToken(hcatToken.getService(), hcatToken);
 
-        // check and see if user specified the extra hcat locations we need to
-        // look at and fetch token.
-        final List<String> extraHcatLocations =
-            props.getStringList(EXTRA_HCAT_LOCATION);
-        if (Collections.EMPTY_LIST != extraHcatLocations) {
-          logger.info("Need to pre-fetch extra metaStore tokens from hive.");
+        // Added support for extra_hcat_clusters LIHADOOP-29840
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Remove references to internal jiras

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 4]
```diff
@@ -525,24 +527,42 @@ public synchronized void prefetchToken(final File tokenFile,
...
-            hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, thriftUrl);
-            hcatToken =
-                fetchHcatToken(userToProxy, hiveConf, thriftUrl, logger);
+            hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, thriftUrls);
+            hcatToken = fetchHcatToken(userToProxy, hiveConf, thriftUrls, logger);
             cred.addToken(hcatToken.getService(), hcatToken);
           }
-
+        } else {
+          // check and see if user specified the extra hcat locations we need to
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Clarify (explicitly) that _LOCATIONS conf is processed only if _CLUSTERS is not provided.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 3 above)
Comment (herrlich) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Sorry forgot that

File: az-core/src/main/java/azkaban/utils/Props.java (hunk 1 above)
Comment (kunkun-tang) [reply az-core/src/main/java/azkaban/utils/Props.java]: +1 on Unit Test! You might need to create a test file under az-core module, as we don't have one.

File: az-core/src/test/java/azkaban/utils/PropsTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2014 LinkedIn Corp.
```
Comment (HappyRay) [inline az-core/src/test/java/azkaban/utils/PropsTest.java]: See the dev guide for how to update the copyright automatically.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 6]
```diff
@@ -121,7 +121,17 @@
...
+  @Deprecated
   private static final String EXTRA_HCAT_LOCATION = "other_hcat_location";
+  /**
+   * this parameter is used to replace EXTRA_HCAT_LOCATION that could fail when one of the uris is not available.
+   * EXTRA_HCAT_CLUSTERS has the following format:
+   * other_hcat_clusters = "(thrift://hcat1:port,thrift://hcat2:port),(thrift://hcat3:port,thrift://hcat4:port)"
+   * Each string in the parenthesis is regarded as a "cluster", which we traverse until we get a delegation token
+   * The uris(hcat servers) in a "cluster" ensures HA is provided.
+   */
+  private static final String EXTRA_HCAT_CLUSTERS = "other_hcat_clusters";
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Please move the keys to the Constant class. See the dev guide.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 7]
```diff
@@ -525,24 +536,43 @@ public synchronized void prefetchToken(final File tokenFile,
...
-            hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, thriftUrl);
-            hcatToken =
-                fetchHcatToken(userToProxy, hiveConf, thriftUrl, logger);
+            hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, thriftUrls);
+            hcatToken = fetchHcatToken(userToProxy, hiveConf, thriftUrls, logger);
             cred.addToken(hcatToken.getService(), hcatToken);
           }
-
+        } else {
+          // If EXTRA_HCAT_CLUSTERS is not specified, we will continue with EXTRA_HCAT_LOCATION
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Update it to say "Only if EXTRA_HCAT_CLUSTERS"

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 8]
```diff
@@ -121,7 +121,17 @@
...
    * supported, use comma to separate the values, values are case insensitive.
    */
+  // Use EXTRA_HCAT_CLUSTERS instead
+  @Deprecated
   private static final String EXTRA_HCAT_LOCATION = "other_hcat_location";
+  /**
+   * this parameter is used to replace EXTRA_HCAT_LOCATION that could fail when one of the uris is not available.
+   * EXTRA_HCAT_CLUSTERS has the following format:
+   * other_hcat_clusters = "(thrift://hcat1:port,thrift://hcat2:port),(thrift://hcat3:port,thrift://hcat4:port)"
+   * Each string in the parenthesis is regarded as a "cluster", which we traverse until we get a delegation token
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: The comment is misleading; the code should not stop when it gets a delegation token from one cluster; it should get delegation tokens from all clusters.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 9]
```diff
@@ -121,7 +121,17 @@
    * default one the system should pre-fetch hcat token from. Note: Multiple thrift uris are
    * supported, use comma to separate the values, values are case insensitive.
    */
+  // Use EXTRA_HCAT_CLUSTERS instead
```
Comment (beeramsunitha) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Move the comment about using clusters into the block comment above.

File: az-core/src/test/java/azkaban/utils/PropsTest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+public class PropsTest {
+  Props p = new Props();
+  private static final String EXTRA_HCAT_CLUSTERS = "other_hcat_clusters";
+
+  /* Test for getStringListFromCluster(String s)*/
+  @Test
+  public void testSplit1() {
+    String s1 = "(thrift://hcat1:port,thrift://hcat2:port),(thrift://hcat3:port,thrift://hcat4:port)";
+    p.put(EXTRA_HCAT_CLUSTERS, s1);
+    List<String> s2 = Arrays.asList("", "thrift://hcat1:port,thrift://hcat2:port" , "thrift://hcat3:port,thrift://hcat4:port");
```
Comment (beeramsunitha) [inline az-core/src/test/java/azkaban/utils/PropsTest.java]: We shouldn't be returning an empty string. The list should only  have 2 entries...

File: az-core/src/test/java/azkaban/utils/PropsTest.java (hunk 5 above)
Comment (herrlich) [reply az-core/src/test/java/azkaban/utils/PropsTest.java]: Copied the unit test template from another file, will modify that.

File: az-core/src/main/java/azkaban/Constants.java [hunk 11]
```diff
@@ -167,6 +167,15 @@
...
     public static final String AZKABAN_JOB_LOGGING_KAFKA_ENABLE = "azkaban.job.logging.kafka.enable";
+
+    /*
+     * this parameter is used to replace EXTRA_HCAT_LOCATION that could fail when one of the uris is not available.
+     * EXTRA_HCAT_CLUSTERS has the following format:
+     * other_hcat_clusters = "thrift://hcat1:port,thrift://hcat2:port;thrift://hcat3:port,thrift://hcat4:port"
+     * Each string in the parenthesis is regarded as a "cluster", and we will get a delegation token from each cluster.
+     * The uris(hcat servers) in a "cluster" ensures HA is provided.
+     **/
+    public static final String EXTRA_HCAT_CLUSTERS = "other_hcat_clusters";
```
Comment (kunkun-tang) [inline az-core/src/main/java/azkaban/Constants.java]: Should this property be put under `JobProperties` class. Also, you might want to use `azkaban.job.other_hcat_clusters`

File: az-core/src/main/java/azkaban/utils/Props.java [hunk 12]
```diff
@@ -430,6 +431,23 @@ public String getString(final String key) {
   }
 
   /**
+   * Returns a list of clusters with the comma as the separator of the value
+   * e.g., for input string: "thrift://hcat1:port,thrift://hcat2:port;thrift://hcat3:port,thrift://hcat4:port;"
```
Comment (kunkun-tang) [inline az-core/src/main/java/azkaban/utils/Props.java]: What if the string has some format error? What will happen? Like 
```
thrift:hcat1:port;thrift:hcat2:port
```

File: az-core/src/test/java/azkaban/utils/PropsTest.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    String s3 = "thrift://hcat1:port,thrift://hcat2:port     ;      thrift://hcat3:port,thrift://hcat4:port;";
+    p.put(EXTRA_HCAT_CLUSTERS, s3);
+    List<String> s4 = Arrays.asList( "thrift://hcat1:port,thrift://hcat2:port" , "thrift://hcat3:port,thrift://hcat4:port");
+    Assert.assertTrue(p.getStringListFromCluster(EXTRA_HCAT_CLUSTERS).equals(s4));
+
+    String s5 = "thrift://hcat1:port,thrift://hcat2:port";
+    p.put(EXTRA_HCAT_CLUSTERS, s5);
+    List<String> s6 = Arrays.asList("thrift://hcat1:port,thrift://hcat2:port");
+    Assert.assertTrue(p.getStringListFromCluster(EXTRA_HCAT_CLUSTERS).equals(s6));
```
Comment (kunkun-tang) [inline az-core/src/test/java/azkaban/utils/PropsTest.java]: Is there any assertFalse cases?

File: az-core/src/main/java/azkaban/Constants.java (hunk 11 above)
Comment (kunkun-tang) [reply az-core/src/main/java/azkaban/Constants.java]: Or, is `other_hcat_cluster`s an existing configuration now?

Comment (HappyRay) [reply az-core/src/main/java/azkaban/Constants.java]: How about azkaban.job.hive.other_hcat_clusters?
Please move the existing config to this class too while you are refactoring.

**Comment by HappyRay** [comment]
Please describe in more details why this config is needed? The relationship with the existing config and the plan for the existing config.

File: az-core/src/main/java/azkaban/utils/Props.java (hunk 12 above)
Comment (herrlich) [reply az-core/src/main/java/azkaban/utils/Props.java]: In this way, I think hive will throw an exception.

File: az-core/src/test/java/azkaban/utils/PropsTest.java (hunk 13 above)
Comment (herrlich) [reply az-core/src/test/java/azkaban/utils/PropsTest.java]: Discussed offline

**Pull request merged**